query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,529 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::btree_map::Entry;
+
+use mentat_core::Schema;
+
+use mentat_query::{
+    NotJoin, 
+    UnifyVars, 
+    ContainsVariables,
+};
+
+use clauses::ConjoiningClauses;
+
+use errors::{
+    Result,
+    ErrorKind,
+};
+
+use types::{
+    ColumnConstraint, 
+    ComputedTable,
+};
+
+impl ConjoiningClauses {
+    pub fn apply_not_join(&mut self, schema: &Schema, not_join: NotJoin) -> Result<()> {
+        let unified = match not_join.unify_vars {
+            UnifyVars::Implicit => not_join.collect_mentioned_variables(),
+            UnifyVars::Explicit(vs) => vs.into_iter().collect(),
+        };
+
+        let mut template = self.use_as_template(&unified);
+        
+        for v in unified.iter() {
+            if self.value_bindings.contains_key(&v) {
+                let val = self.value_bindings.get(&v).unwrap().clone();
+                template.value_bindings.insert(v.clone(), val);
+            } else if !self.column_bindings.contains_key(&v) {
+                bail!(ErrorKind::UnboundVariable(v.name()));
+            }
+        }
+
+        for clause in not_join.clauses.into_iter() {
+            template.apply_clause(&schema, clause)?;
+        }
+
+        if template.is_known_empty() {
+            return Ok(());
+        }
+
+        for v in unified.iter() {
+           if self.column_bindings.contains_key(&v) {
```

Comment:
When I put that in 46, things started going a bit wrong.

For
```
            [:find ?x ?age
             :where
             [?x :foo/age ?age]
             [[< ?age 30]]
             (not [?x :foo/knows "John"]
                  [?x :foo/knows "Daphne"])]
```
The output stopped joining `datoms01.e` with `datoms02.e` inside the subquery, but joined `datoms02.e` with `datoms00.e` instead, meaning both `datoms01.e` and `datoms02.e` are joined with `datoms00.e`, which strikes me as wrong. 

New Output:
```
Constraint(datoms00.a = entity(68)), Constraint(datoms00.v < value(Long(30))), Constraint(NOT EXISTS Subquery(ConjoiningClauses { empty_because: None, from: [SourceAlias(Datoms, datoms01), SourceAlias(Datoms, datoms02)], wheres: [Constraint(datoms01.a = entity(66)), Constraint(datoms01.v = value(String("John"))), Constraint(datoms02.a = entity(66)), Constraint(datoms02.v = value(String("Daphne"))), Constraint(datoms00.e = datoms01.e), Constraint(datoms00.e = datoms02.e)], column_bindings: {var(?x): [datoms00.e, datoms01.e, datoms02.e]}, input_variables: {}, value_bindings: {}, known_types: {var(?x): ValueTypeSet({Ref})}, extracted_types: {} }))
```

Expected Output:
```
Constraint(datoms00.a = entity(68)), Constraint(datoms00.v < value(Long(30))), Constraint(NOT EXISTS Subquery(ConjoiningClauses { empty_because: None, from: [SourceAlias(Datoms, datoms01), SourceAlias(Datoms, datoms02)], wheres: [Constraint(datoms01.a = entity(66)), Constraint(datoms01.v = value(String("John"))), Constraint(datoms02.a = entity(66)), Constraint(datoms02.v = value(String("Daphne"))), Constraint(datoms00.e = datoms01.e)], Constraint(datoms01.e = datoms02.e), column_bindings: {var(?x): [datoms00.e, datoms01.e, datoms02.e]}, input_variables: {}, value_bindings: {}, known_types: {var(?x): ValueTypeSet({Ref})}, extracted_types: {} }))
```